### PR TITLE
[Clang] Emit DW_TAG_template_alias for template aliases

### PR DESCRIPTION
**Pull request by OCHyams**
Fix issue https://github.com/llvm/llvm-project/issues/54624

Add front end flags -gtemplate-alias (also a cc1 flag) and -gno-template-alias to enable/disable usage of the feature. It's enabled by default in the front end for SCE debugger tuning only. GCC emits DW_TAG_typedef for template aliases, and GDB and LLDB appear not to support DW_TAG_template_alias.

The -Xclang option -gsimple-template-names=mangled is treated the same as =full, which is not a regression from current behaviour for type aliases.

Use DICompositeType to represnt the template alias, using its extraData field as a tuple of DITemplateParameter to describe the template parameters.

New tests:
```
  template-alias.cpp - Check the metadata construction & interaction with
                       -gsimple-template-names.
  template-alias.ll  - Check DWARF emission.
```

Modified tests:
```
  debug-options.c    - Check Clang generates correct cc1 flags.
  frame-types.s      - Check llvm-symbolizer understands the DIE.
  ```

### CODE REVIEW

File: core
Comment (pogo59) [inline core:1]: oops

File: llvm/include/llvm/IR/DIBuilder.h
Comment (pogo59) [inline llvm/include/llvm/IR/DIBuilder.h:321]: (optional)

File: llvm/lib/CodeGen/AsmPrinter/DebugHandlerBase.cpp
Comment (pogo59) [inline llvm/lib/CodeGen/AsmPrinter/DebugHandlerBase.cpp:159]: Looks like this got added twice?

File: clang/lib/CodeGen/CGDebugInfo.cpp
Comment (pogo59) [inline clang/lib/CodeGen/CGDebugInfo.cpp:5395]: The sentence inside the parens should be a proper sentence.

**Review by pogo59 — COMMENTED** [review]
Quick initial pass.

Converting frame-types.s from v4 to v5 should be done as a separate commit. Then adding the alias would be a simpler change to review. (Is symbolizer really the right place to be testing this, anyway?)

I'm a little uncomfortable with adding a new user-facing option for template aliases. Even with that in place, we should not warn and refuse to do what the user asked for, based on DWARF version. `-gdwarf-2 -gsplit-dwarf` generates a .dwo file claiming to be v2, for example.

**Comment by OCHyams** [comment]
> Quick initial pass.

Much appreciated!

> Converting frame-types.s from v4 to v5 should be done as a separate commit.  Then adding the alias would be a simpler change to review.

Makes sense, I can do that.

> (Is symbolizer really the right place to be testing this, anyway?)

I modified frame-types.s here because  `getTypeSizeImpl` in `DWARFDie.cpp` is stimulated by that test, and I've made a change to that function in this patch.

> I'm a little uncomfortable with adding a new user-facing option for template aliases. Even with that in place, we should not warn and refuse to do what the user asked for, based on DWARF version. -gdwarf-2 -gsplit-dwarf generates a .dwo file claiming to be v2, for example.

I don't have strong feelings about the flag setup and happy to change it. @dwblaikie mentioned that we might not want to have the debugger tuning be the only way of controlling whether or not we emit this DIE, so adding `-gtemplate-alias` which has its default set based on debugger tuning seemed like a reasonable way of doing it. What would be the best way of controlling this feature in your opinion (bearing in mind that GCC doesn't emit it, and GDB and LLDB don't the DIE)?

File: clang/lib/CodeGen/CGDebugInfo.cpp
Comment (OCHyams) [reply clang/lib/CodeGen/CGDebugInfo.cpp:5395]: These are existing comments that I've just translocated - this block has been copied from line 5390, below. I'm happy to make your suggested change, I just wanted to add this context first in case it changes your view?

File: llvm/test/DebugInfo/X86/template-alias.ll
Comment (adrian-prantl) [inline llvm/test/DebugInfo/X86/template-alias.ll:43]: x86_64-unknown-unkown

**Review by adrian-prantl — COMMENTED** [review]
LLVM IR parts look good to me.

**Comment by OCHyams** [comment]
> LLVM IR parts look good to me.

Thanks!

@pogo59 said:

> Converting frame-types.s from v4 to v5 should be done as a separate commit. Then adding the alias would be a simpler change to review.

I've split that into two commits in this PR: upgrade to v5: [18446f2](https://github.com/llvm/llvm-project/pull/87623/commits/18446f27e31803057d9d90c957e5e191eb263b7b), add template alias: [49d6642](https://github.com/llvm/llvm-project/pull/87623/commits/49d66423ae61baf9cacd42c2b57333e09e4c1e81). It does improve readability a bit but its still quite noisy. I am happy to commit the former separately then rebase this PR, if you'd like?

File: clang/test/CodeGenCXX/debug-info-alias.cpp
Comment (Michael137) [inline clang/test/CodeGenCXX/debug-info-alias.cpp:1]: Do we now lose a tiny bit of coverage for `-glldb`? I assume this change is to make sure the test passes on platforms that are tuned for `sce`? Maybe we should just add `-gno-template-alias` instead?

File: clang/lib/CodeGen/CGDebugInfo.cpp
Comment (Michael137) [inline clang/lib/CodeGen/CGDebugInfo.cpp:5435]: Are we using this extra `Ty` parameter anywhere? I assume your original intent was to pass the `TemplateSpecializationType` into `GetName` and get the template arguments for the alias decl that way? Any reason you didn't end up doing that?

**Comment by dwblaikie** [comment]
> > I'm a little uncomfortable with adding a new user-facing option for template aliases. Even with that in place, we should not warn and refuse to do what the user asked for, based on DWARF version. -gdwarf-2 -gsplit-dwarf generates a .dwo file claiming to be v2, for example.
> 
> I don't have strong feelings about the flag setup and am happy to change it. @dwblaikie mentioned that we might not want to have the debugger tuning be the only way of controlling whether or not we emit this DIE, so adding `-gtemplate-alias` which has its default set based on debugger tuning seemed like a reasonable way of doing it. What would be the best way of controlling this feature in your opinion (bearing in mind that GCC doesn't emit it, and GDB and LLDB don't the DIE)?

@pogo59 Hmm - I thought you held a strong preference that debugger tuning never be the only way to access a certain feature, and that we should always have direct control of these features via flags (but with default (or explicit) debugger tuning setting some of these flag defaults based on the preferred content for that debugger). Perhaps i'm misremembering?

**Comment by dwblaikie** [comment]
Re: code: Looks about right.

Bit unfortunate to store template parameters in different ways (in the `extraData` for the alias template, but in the `templateParams` for the composite types) - but I guess it'd be more invasive to try to represent alias templates as composite types?

File: clang/lib/CodeGen/CGDebugInfo.cpp
Comment (Michael137) [reply clang/lib/CodeGen/CGDebugInfo.cpp:5435]: Ah thanks for the additional context. Keeping it as is seems fine to me then. I don't see a better way to get to template arguments from the alias decl. My initial thought of special-casing this in `GetName` isn't necessarily great either

**Comment by pogo59** [comment]
> @pogo59 Hmm - I thought you held a strong preference that debugger tuning never be the only way to access a certain feature, and that we should always have direct control of these features via flags (but with default (or explicit) debugger tuning setting some of these flag defaults based on the preferred content for that debugger). Perhaps i'm misremembering?

You are not misremembering. Ideally _tuning_ is not _gating_ (although we probably never expressed it that way). We probably have not fully held to that ideal, but if we're going to have knobs to tweak we should implement the knobs properly. I will be making another pass over this PR later today and pay attention to that.

**Comment by OCHyams** [comment]
Thanks @dwblaikie 

> Bit unfortunate to store template parameters in different ways (in the extraData for the alias template, but in the templateParams for the composite types) - but I guess it'd be more invasive to try to represent alias templates as composite types?

I hadn't actually tried using DICompsiteType. I didn't choose DIDerivedType for any particularly principled reasons: typedefs use derived types which made it very easy to find places that needed updating, and an alias "felt" more like a derived type than composite type. Having looked only briefly, I don't _think_ it'd be any more invasive to use composite types instead (not 100% sure without diving in). I'm happy to give that a go if that is your preference?

---

I've just found an input that causes an assertion failure in `CollectTemplateParams` with my implementation:
```
template<typename Y, typename Z>
struct X {
  Y m1;
  Z m2;
};

template<typename... Ts>
using A = X<Ts...>;

A<int, int> a;
```

There's a mismatch between the template parameter list and template argument list sizes, created here:
```
  if (CGM.getCodeGenOpts().DebugTemplateAlias) {
    TemplateArgs Args = {TD->getTemplateParameters(), Ty->template_arguments()}; // <--- here
 ```
 
I notice that  we emit `DW_TAG_GNU_template_parameter_pack` for, e.g., variadic template struct instantiations. I've not figured out how to fix this just yet, but thought I'd bring it up in case there's something obvious.

**Comment by dwblaikie** [comment]
> Thanks @dwblaikie
> 
> > Bit unfortunate to store template parameters in different ways (in the extraData for the alias template, but in the templateParams for the composite types) - but I guess it'd be more invasive to try to represent alias templates as composite types?
> 
> I hadn't actually tried using DICompsiteType. I didn't choose DIDerivedType for any particularly principled reasons: typedefs use derived types which made it very easy to find places that needed updating, and an alias "felt" more like a derived type than composite type. Having looked only briefly, I don't _think_ it'd be any more invasive to use composite types instead (not 100% sure without diving in). I'm happy to give that a go if that is your preference?

Eh, I guess if we've already got DIGlobalVariable having template params, we aren't going to unify all 3 cases - not sure if unifying 2 out of three is particularly valuable, especially if it ends up with a different conflict (that alias templates look less like aliases & so need special casing in that direction)

But perhaps at least you could add named accessors to DIDerivedType for the template params, same as DIGlobalVariable/DICompositeType have?

(oh, and in case anyone hasn't mentioned it already - this would generally be committed in smaller pieces upstream - adding the LLVM functionality first, then adding clang patches that use that functionality)

> I've just found an input that causes an assertion failure in `CollectTemplateParams` with my implementation:
> 
> ```
> template<typename Y, typename Z>
> struct X {
>   Y m1;
>   Z m2;
> };
> 
> template<typename... Ts>
> using A = X<Ts...>;
> 
> A<int, int> a;
> ```
> 
> There's a mismatch between the template parameter list and template argument list sizes, created here:
> 
> ```
>   if (CGM.getCodeGenOpts().DebugTemplateAlias) {
>     TemplateArgs Args = {TD->getTemplateParameters(), Ty->template_arguments()}; // <--- here
> ```
> 
> I notice that we emit `DW_TAG_GNU_template_parameter_pack` for, e.g., variadic template struct instantiations. I've not figured out how to fix this just yet, but thought I'd bring it up in case there's something obvious.

yeah, probably worth checking the APIs that we use for building the template args for classes and variable templates, they probably have some handling for this - would be good to share/reuse that code, rather than duplicating it, once you find it

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (pogo59) [inline clang/lib/Driver/ToolChains/Clang.cpp:4609]:
```suggestion
    else
```

File: clang/lib/Driver/ToolChains/Clang.cpp
Comment (pogo59) [inline clang/lib/Driver/ToolChains/Clang.cpp:4611]: Please take out the diags here, if someone explicitly asks for something in debug info we do it.

File: clang/lib/CodeGen/CGDebugInfo.cpp
Comment (pogo59) [reply clang/lib/CodeGen/CGDebugInfo.cpp:5395]: Ah, didn't notice it was a copy-paste. But it would still be nice to have the comment follow our style guide. So, up to you.

File: clang/test/Driver/debug-options.c
Comment (pogo59) [inline clang/test/Driver/debug-options.c:470]: `--target`

**Comment by OCHyams** [comment]
> But perhaps at least you could add named accessors to DIDerivedType for the template params, same as DIGlobalVariable/DICompositeType have?

Done. I couldn't add an assert (`getTag() == dwarf::DW_TAG_template_alias`) without including `Dwarf.h`. I've chosen to not do that as it looks like its been avoided (but possibly just not really needed here before), but I'd be happy to add it if the extra compile time isn't worth worrying about.

> (oh, and in case anyone hasn't mentioned it already - this would generally be committed in smaller pieces upstream - adding the LLVM functionality first, then adding clang patches that use that functionality)

Sure, no problem, I'll open separate pull requests once this has been accepted then.

---

I think I've addressed all the concerns raised now except for the variadic issue I mentioned, which I'll look at tomorrow.

**Comment by OCHyams** [comment]
> yeah, probably worth checking the APIs that we use for building the template args for classes and variable templates, they probably have some handling for this - would be good to share/reuse that code, rather than duplicating it, once you find it

I didn't find anything that does what (I think) we need here (but I'm not familiar with the FE, so might've just missed it if it exists. Not for lack of trying). The code I've added looks a bit ugly, but I'm not sure what else to do. See https://godbolt.org/z/nTK4bW1Yn - there doesn't seem to be a `TemplateArgument pack` for the template alias `ClassTemplateSpecializationDecl` (while there is one for the struct `E` in that link), this this is what I came up with instead.

@dwblaikie are you happy with the latest addition mentioned above (plus test) - that's the commits [dc5a611](https://github.com/llvm/llvm-project/pull/87623/commits/dc5a6116f40fd7b6eb4572c25594cd82f3583208) and [c9c000f](https://github.com/llvm/llvm-project/pull/87623/commits/c9c000fb30d629b29b39f6ffd8a25028a0c00999)?

@pogo59 are you happy with the flag situation now?

File: clang/lib/CodeGen/CGDebugInfo.cpp
Comment (pogo59) [inline clang/lib/CodeGen/CGDebugInfo.cpp:1339]: substituted

File: clang/test/Driver/debug-options.c
Comment (pogo59) [reply clang/test/Driver/debug-options.c:470]: Huh Coulda sworn that was changed... okay.

File: llvm/include/llvm/IR/DIBuilder.h
Comment (pogo59) [inline llvm/include/llvm/IR/DIBuilder.h:322]: Flags is also (optional)

**Review by pogo59 — COMMENTED** [review]
The flag situation is good. Tuning influences the default for the new option, which is how we want that to work.

I'm happy, a couple of nits left. I'll leave it to others to do the final approval.

File: clang/lib/CodeGen/CGDebugInfo.cpp
Comment (Michael137) [inline clang/lib/CodeGen/CGDebugInfo.cpp:1356]: Does this mean we won't be emitting the defaulted template arguments as part of the `DW_TAG_template_alias`? Would be good to add a test case for defaulted arguments, so this logic is exercised (if you haven't already)

Also minor nit, is this logic getting sufficiently long to warrant a separate helper function?

**Comment by Michael137** [comment]
Btw, as a follow-up to this patch should we check that this is compatible with dsymutil (i.e., running `dsymutil --verify`)? I suspect it might need a fixup (given LLDB doesn't even support this tag)

**Review by OCHyams — COMMENTED** [review]
(oops, didn't submit my inline replies...)

File: clang/lib/CodeGen/CGDebugInfo.cpp
Comment (Michael137) [reply clang/lib/CodeGen/CGDebugInfo.cpp:1356]: Ah I see, thanks for clarifying! Just thought I'd point it out in case this wasn't expected. The DWARF spec has this to say about the template parameters of an alias:
```
The template alias entry has child entries describing the template actual parameters (see Section 2.23 on page 57).
```
Which to me implies that the parameters of an alias instantiation follow the same rules as those for class template instantiations.

Regarding name reconstruction, defaulted template arguments should have their `IsDefaulted` bit set, so when reconstructing the name we should be able to drop defaulted arguments? If that's too much of a hassle maybe not emitting them at all as you do now is also fine?

Comment (OCHyams) [reply clang/lib/CodeGen/CGDebugInfo.cpp]: I'm glad you pointed it out - it was doing what I wanted it to but I hadn't tried very hard to get it to gather default argument values, and I agree copying the class template instantiation debug info is the safest bet.

I've implemented that now. Slightly worried there are other template parameter kinds I've not thought of, currently guarded by an llvm_unrecahable. Maybe there's something better to do there - do you have any thoughts on this?

Comment (OCHyams) [reply clang/lib/CodeGen/CGDebugInfo.cpp]: Ah - this implementation appears to cause an assertion when one of the template parameters is a dependant type... (working on it)

File: llvm/include/llvm/IR/DIBuilder.h
Comment (pogo59) [reply llvm/include/llvm/IR/DIBuilder.h:322]: Ah sorry missed that.

File: clang/lib/CodeGen/CGDebugInfo.cpp
Comment (Michael137) [reply clang/lib/CodeGen/CGDebugInfo.cpp]: > lightly worried there are other template parameter kinds I've not thought of, currently guarded by an llvm_unrecahable. Maybe there's something better to do there - do you have any thoughts on this?

The three kinds you check for should be sufficient afaik. Places around clang consistently only deal with these three types of parameter decls.

File: clang/lib/CodeGen/CGDebugInfo.cpp
Comment (Michael137) [inline clang/lib/CodeGen/CGDebugInfo.cpp:1362]: Could we get away with just doing `SpecArgs.push_back(P->getDefaultArgument())`? Or were there some lifetime concerns?

Otherwise LGTM

File: clang/lib/CodeGen/CGDebugInfo.cpp
Comment (OCHyams) [reply clang/lib/CodeGen/CGDebugInfo.cpp]: > The three kinds you check for should be sufficient afaik. Places around clang consistently only deal with these three types of parameter decls.

Excellent, that is reassuring, thanks.

I'm wondering if you've got any ideas about my dependent value situation:

```
template <int I1 = 5, int I2 = I1>
using B = int;
B<> b;
```

The current default-argument-gathering tactic can't cope with dependent default expressions/types. Do you know if there's any way that I can "resolve", for lack of a better word, the template parameter default expression to get an expression `= 5` rather than the dependent `= I1`? Presumably this is doable, because we've got a specialisation that defines what `I1` is (are there cases where this could never work... maybe with SFINAE? I'm not sure).

I'm continuing to look into it, I'm only asking in case you know off the top of your head, no problem if not.

Comment (Michael137) [reply clang/lib/CodeGen/CGDebugInfo.cpp]: Clang doesn't resolve defaultedness for dependent arguments either. As a consequence, when a template parameter has a dependent default argument then its `DW_TAG_template_value_parameter` won't have a `DW_AT_default_value`. So I don't think we need to handle this if it's going to be hard to do. Can we just arrange for the parameter to be emitted as a non-defaulted `DW_TAG_template_value_parameter`?

Comment (OCHyams) [reply clang/lib/CodeGen/CGDebugInfo.cpp]: Looking at a class template instantiation with a default dependent value:
```
template <int I1 = 5, int I2 = I1>
struct B {};
B<> b;
```
We get:
```
0x00000029:   DW_TAG_structure_type
                DW_AT_calling_convention        (DW_CC_pass_by_value)
                DW_AT_name      ("B<5, 5>")
                DW_AT_byte_size (0x01)
                DW_AT_decl_file ("/home/och/scratch/test11.cpp")
                DW_AT_decl_line (2)

0x0000002f:     DW_TAG_template_value_parameter
                  DW_AT_type    (0x0000003e "int")
                  DW_AT_name    ("I1")
                  DW_AT_default_value   (true)
                  DW_AT_const_value     (5)

0x00000036:     DW_TAG_template_value_parameter
                  DW_AT_type    (0x0000003e "int")
                  DW_AT_name    ("I2")
                  DW_AT_default_value   (true)
                  DW_AT_const_value     (5)

0x0000003d:     NULL
 ```
 
Is your suggestion to emit something like this for the template alias case (from my previous comment)?
```
0x00000029:   DW_TAG_template_alias
                DW_AT_name      ("B<5, ?>")
                ...

0x0000002f:     DW_TAG_template_value_parameter
                  DW_AT_type    (0x0000003e "int")
                  DW_AT_name    ("I1")
                  DW_AT_default_value   (true)
                  DW_AT_const_value     (5)

0x00000036:     DW_TAG_template_value_parameter
                  DW_AT_type    (0x0000003e "int")
                  DW_AT_name    ("I2")

0x0000003d:     NULL
```

I think that is doable but I'm not sure how that would interact with name reconstruction (in Clang debug info without -gsimple-template-names, and in debuggers/tools with it).

tyvm for your ongoing help with this.

Comment (Michael137) [reply clang/lib/CodeGen/CGDebugInfo.cpp]: Ah apologies! I must've misremembered, I could've sworn there were cases where we don't handle the defaulted-ness.

> Is your suggestion to emit something like this for the template alias case (from my previous comment)?
Yup, would default values ever matter for name reconstruction? The only thing I know of that uses `DW_AT_default_value` in LLDB is when formatting the typename of a variable when printing it (we omit defaulted template parameters).

Comment (Michael137) [reply clang/lib/CodeGen/CGDebugInfo.cpp]: I suppose you could call `EvaluateKnownConstInt` (on the TemplateArgument expression, if it's `!isValueDependent() && isIntegerConstantExpr()`, similar to what `templateArgumentExpressionsEqual` does). Maybe we can do this from within `CollectTemplateParams` where it asserts that the expression is a Constant?

I'm not actually sure how we get to the `TemplateArgument::Expression` case in `CollectTemplateParams` for constant expressions. AFAICT in the AST such expressions get evaluated into `TemplateArgument::Integral` before we ever get here. Seems like you're tripping over the assert now because you're calling `CollectTemplateParams` on the arguments to a `TemplateAliasDecl` which don't have their arguments evaluated in the front-end it looks like.

Comment (OCHyams) [reply clang/lib/CodeGen/CGDebugInfo.cpp]: (defaulted values... and types*)

Comment (OCHyams) [reply clang/lib/CodeGen/CGDebugInfo.cpp]: > I suppose you could call EvaluateKnownConstInt (on the TemplateArgument expression, if it's !isValueDependent() && isIntegerConstantExpr(), similar to what templateArgumentExpressionsEqual does). Maybe we can do this from within CollectTemplateParams where it asserts that the expression is a Constant?

The constant value default appears to be working, it's the dependent value `int I2 = I1` that is causing problems for me.

> Seems like you're tripping over the assert now because you're calling CollectTemplateParams on the arguments to a TemplateAliasDecl which don't have their arguments evaluated in the front-end it looks like.

That sounds like it might be the problem. It feels like we have all the parts needed to instantiate the dependent expression at this point... (because all the types/values it depends on exist as non-default argument values), I'm just not sure how to ask Clang to do that.


Here's an example using type parameters rather than values (you can also do `typename C = X<B>` if you want to hit yet another assertion...)
```
template <typename B = int, typename C = B>
using A = int;
A<> a;
```

```
Dependent types cannot show up in debug information
UNREACHABLE executed at clang/lib/CodeGen/CGDebugInfo.cpp:3686!
...
CGDebugInfo::CreateTypeNode(clang::QualType, llvm::DIFile*)
CGDebugInfo::getOrCreateType(clang::QualType, llvm::DIFile*) 
CollectTemplateParams # from the TemplateArgument::Type case processing the type of the argument for C
```

Comment (OCHyams) [reply clang/lib/CodeGen/CGDebugInfo.cpp]: >Sorry for diverting you down this path, maybe ignoring the default arguments as you did in your first attempt is sufficient if this is just not feasible.

Not at all, I think it was worth exploring and I very much appreciate the detailed review and the help that's come with it.

I agree that ignoring default arguments might be the way forward then. It isn't a regression in terms of name reconstruction, since the existing typedef approach for template aliases omits defaulted arguments from the DW_AT_name anyway:

```
template <typename B = int, typename C = B>
using A = int;
A<> a;
```
`$ clang test.cpp -g -o - | llvm-dwarfdump -o -`
```
[...]
0x0000002e:   DW_TAG_typedef
                DW_AT_type      (0x00000036 "int")
                DW_AT_name      ("A<>")  <-- no defaulted arguments for template aliases :-(
                DW_AT_decl_file ("test.cpp")
                DW_AT_decl_line (41)
```

I have pushed that change (to omit defaulted arguments). Thanks for bearing with me on this one!

I've chatted to our debugger folks and they're ok with omitting defaulted args (though ideally this would be fixed in the future).

I think moving this discussion onto discourse might be useful - I'll look at doing that later in the week or next week.

Comment (Michael137) [reply clang/lib/CodeGen/CGDebugInfo.cpp]: Sounds good! FWIW, LLDB would currently omit the defaulted parameters too, by choice, since usually they're just noise (especially with many of the STL types).

File: clang/test/CodeGenCXX/defaulted-template-alias.cpp
Comment (Michael137) [inline clang/test/CodeGenCXX/defaulted-template-alias.cpp:33]: Minor: Is it worth turning these into `CHECK-NOT`? So when this is fixed in the future we remember that we have these tests ready

**Review by Michael137 — APPROVED** [review]
LGTM, if @pogo59 and @dwblaikie are happy with the driver changes

File: clang/test/CodeGenCXX/defaulted-template-alias.cpp
Comment (OCHyams) [reply clang/test/CodeGenCXX/defaulted-template-alias.cpp:33]: I think the test will already start failing if this gets fixed because the current `extraData` metadata tuple contents check will fail (it'll have more than one element). I'm still happy to add CHECK-NOTs if you'd like, just thought I'd raise this in case it changes your stance.

Comment (Michael137) [reply clang/test/CodeGenCXX/defaulted-template-alias.cpp:33]: ah right, that makes sense, happy keeping it as is

**Comment by pogo59** [comment]
Yes, driver part LGTM.

**Comment by OCHyams** [comment]
Thanks everyone. I'll land this now then to see what the bots have to say about it.

**Comment by OCHyams** [comment]
Dependent expressions strike again - https://godbolt.org/z/W381837vr
```
template <int>
using A = int;

template<int I>
struct S {
  using AA = A<I>;
  AA aa;
};

S<0> s;
```
` clang++ -c test.cpp -g -gtemplate-alias`
`clang/lib/AST/ExprConstant.cpp:15721: bool clang::Expr::EvaluateAsRValue(clang::Expr::EvalResult&, const clang::ASTContext&, bool) const: Assertion '!isValueDependent() && "Expression evaluator can't be called on a dependent expression."' failed.`

The issue is `I` is a dependent expression. We have an actual instantiation of S with the value of `I` clearly being `0` here, but I'm not sure whether that information exists in the AST for the `TemplateSpecializationType *`.

Looking into this now (will revert if I can't find a solution).

**Comment by OCHyams** [comment]
@Michael137 said:


> Btw, as a follow-up to this patch should we check that this is compatible with dsymutil (i.e., running dsymutil --verify)? I suspect it might need a fixup (given LLDB doesn't even support this tag)

`dsymutil --verify` seems to be happy with the tag. Upon closer inspection, I think it uses the same "verifier" code as llvm-dwarfdump so that makes sense (#89589).

There's a couple of switches in llvm/lib/DWARFLinker that look like they want a DW_TAG_template_alias:
`DependencyTracker::isTypeTableCandidate` in `DependencyTracker.cpp` 
`AcceleratorRecordsSaver::save` in `AcceleratorRecordsSaver.cpp`. I'm not sure what a test for those would look like but I can look into it if you'd like?

**Comment by Michael137** [comment]
> @Michael137 said:
> 
> > Btw, as a follow-up to this patch should we check that this is compatible with dsymutil (i.e., running dsymutil --verify)? I suspect it might need a fixup (given LLDB doesn't even support this tag)

Yup, dsymutil looks good now, thanks

> There's a couple of switches in llvm/lib/DWARFLinker that look like they want a DW_TAG_template_alias:
DependencyTracker::isTypeTableCandidate in DependencyTracker.cpp
AcceleratorRecordsSaver::save in AcceleratorRecordsSaver.cpp. I'm not sure what a test for those would look like but I can look into it if you'd like?

Good question, I'm not familiar with these either, but would make sense to address these eventually. Though I don't think it's urgent atm since `-gtemplate-alias` isn't the default

**Pull request merged**
